lib/pages/class_feedback/service/feedback_provider.dart

```diff
@@ -0,0 +1,147 @@
+    final Map<String, dynamic> data = {};
+
+    if (questionAnswer != null) data['answer'] = questionAnswer;
+    data['dateSubmitted'] = Timestamp.now();
+    data['class'] = className;
+    data['teacher'] = teacher.name;
+    data['assistant'] = assistant.name;
+
+    return data;
+  }
+}
+
+extension FeedbackQuestionExtension on FeedbackQuestion {
+  static FeedbackQuestion fromJSON(dynamic json, String id) {
+    if (json['type'] == 'dropdown' && json['options'] != null) {
+      final List<dynamic> options = json['options'];
+      final List<String> optionsString =
+          options.map((e) => e[LocaleProvider.localeString] as String).toList();
+      return FeedbackQuestionDropdown(
+        category: json['category'],
+        question: json['question'][LocaleProvider.localeString],
+        id: id,
+        answerOptions: optionsString,
+      );
+    } else if (json['type'] == 'rating') {
+      return FeedbackQuestionRating(
+        category: json['category'],
+        question: json['question'][LocaleProvider.localeString],
+        id: id,
+      );
+    } else if (json['type'] == 'text') {
+      return FeedbackQuestionText(
+        category: json['category'],
+        question: json['question'][LocaleProvider.localeString],
+        id: id,
+      );
+    } else if (json['type'] == 'slider') {
+      return FeedbackQuestionSlider(
+        category: json['category'],
+        question: json['question'][LocaleProvider.localeString],
+        id: id,
+      );
+    } else {
+      return FeedbackQuestion(
+        category: json['category'],
+        question: json['question'][LocaleProvider.localeString],
+        id: id,
+      );
+    }
+  }
+}
+
+class FeedbackProvider with ChangeNotifier {
+  Future<bool> addResponse(FeedbackQuestionAnswer response) async {
+    try {
+      await FirebaseFirestore.instance
+          .collection('forms')
+          .doc('class_feedback_answers')
+          .collection(response.questionNumber)
+          .add(response.toData());
+      return true;
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return false;
+    }
+  }
+
+  Future<Map<String, FeedbackQuestion>> fetchQuestions() async {
+    try {
+      final DocumentSnapshot documentSnapshot = await FirebaseFirestore.instance
+          .collection('forms')
+          .doc('class_feedback_questions')
+          .get();
+      final Map<String, dynamic> data = documentSnapshot['questions'];
+      final Map<String, FeedbackQuestion> questions = {};
+      for (final value in data.values) {
+        final key = data.keys.firstWhere((element) => data[element] == value);
+        questions[key] = FeedbackQuestionExtension.fromJSON(value, key);
+      }
+      return questions;
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return null;
+    }
+  }
+
+  Future<Map<String, dynamic>> fetchCategories() async {
```

Comment:
Why `dynamic`? Don't we know it's a `Map<String, String>`? (So the return value here would be `Map<Map<String, String>>`)

Maybe write in a comment what the return value represents here, since it's not obvious.